### PR TITLE
Azure distributed training

### PR DESCRIPTION
**Pull request by marrrcin**
Adds support for distributed training using native Azure ML capabilities.

### CODE REVIEW

File: kedro_azureml/cli_functions.py [hunk 1]
```diff
@@ -26,17 +28,40 @@ def get_context_and_pipeline(ctx: CliContext, image: str, pipeline: str, params)
...
         yield mgr, az_pipeline
 
 
-def parse_extra_params(params):
+def parse_extra_params(params, silent=False):
     if params and (parameters := json.loads(params.strip("'"))):
-        click.echo(
-            f"Running with extra parameters:\n{json.dumps(parameters, indent=4)}"
-        )
+        if not silent:
```
Comment (Lasica) [inline kedro_azureml/cli_functions.py:45]: consider reading silent from params

File: tests/test_distributed.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+        [
+            node(identity, inputs="input_data", outputs="i2", name="node1"),
+            node(my_distributed_node, inputs="i2", outputs="i3", name="node2"),
+            node(identity, inputs="i3", outputs="output_data", name="node3"),
+        ]
+    )
+
+    assert p is not None and hasattr(
+        p.nodes[1].func, DISTRIBUTED_CONFIG_FIELD
+    ), "Distributed note was not marked properly"
```
Comment (Lasica) [inline tests/test_distributed.py:31]: node or note?

File: tests/test_distributed.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    p = pipeline(
+        [
+            node(identity, inputs="input_data", outputs="i2", name="node1"),
+            node(my_distributed_node, inputs="i2", outputs="i3", name="node2"),
+            node(identity, inputs="i3", outputs="output_data", name="node3"),
+        ]
+    )
+
+    assert p is not None and hasattr(
+        p.nodes[1].func, DISTRIBUTED_CONFIG_FIELD
```
Comment (Lasica) [inline tests/test_distributed.py:30]: maybe add assert that not decorated node (like node 0) is not tagged as well.

**Review by Lasica — COMMENTED** [review]
LGTM, Left some comments for potential improvements.
Main thing that's lacking for me is some sort of README or example usage update of this feature somewhere.
Also - update the changelog :)

File: kedro_azureml/distributed/config.py [hunk 4]
```diff
@@ -0,0 +1,22 @@
+import json
+from dataclasses import asdict, dataclass
+from typing import Optional, Union
+
+
+class Framework:
```
[the remark this answers is not in the document]
Comment (marrrcin) [reply kedro_azureml/distributed/config.py:6]: I was considering it, but I was getting `Object of type Framework is not JSON serializable` on the usages of downstream `DistributedNodeConfig` class serialization, so I've just left it as-is with static string fields.

File: kedro_azureml/cli_functions.py [hunk 5]
```diff
@@ -26,17 +28,40 @@ def get_context_and_pipeline(ctx: CliContext, image: str, pipeline: str, params)
...
     return parameters
+
+
+def is_distributed_master_node() -> bool:
+    is_rank_0 = True
+    if "TF_CONFIG" in os.environ:
+        # TensorFlow
+        tf_config = json.loads(os.environ["TF_CONFIG"])
+        worker_type = tf_config["task"]["type"].lower()
+        is_rank_0 = worker_type == "chief" or worker_type == "master"
```
[the remark this answers is not in the document]
Comment (marrrcin) [reply kedro_azureml/cli_functions.py]: Nice catch!

File: kedro_azureml/generator.py [hunk 6]
```diff
@@ -81,9 +98,8 @@ def get_target_resource_from_node_tags(self, node: Node) -> str:
         )
         if len(resource_tags) > 1:
             raise ConfigException(
-                (
-                    "Node tags contain two values that are in defined in the resource config,"
-                )("a node can only have a maximum of 1 resource")
```
Comment (szczeles) [inline kedro_azureml/generator.py:86]: :o

Comment (marrrcin) [reply kedro_azureml/generator.py:86]: #20 , I don't know

**Pull request merged**